### PR TITLE
Bug 1890038: update AWS platform status migration controller for all platforms

### PR DESCRIPTION
**Pull request by staebler**
The AWS platform status migration controller was created to populate status.platformStatus in the infrastructure resource for AWS clusters created in 4.1, prior to the existence of that field. One of the fields that the controller sets is the status.platformStatus.type field, which is set to the type of the platform--in this case AWS. Other operators are looking to that field to determine which platform the cluster is running on. However, for non-AWS clusters created in 4.1, the field remains blank.

With these changes, the AWS platform status migration controller has been re-purposed to fill out the appropriate parts of status.platformStatus for all platforms, not just for AWS. The only other platform used in 4.1 besides AWS was None, and there are no fields in the infrastructure resource. Therefore, for all platforms other than AWS, the only field that the controller populates is the status.platformStatus.type field.

### CODE REVIEW

**Comment by staebler** [comment]
/cc @jstuever @patrickdillon 

Would one of you mind reviewing this PR, please?

**Comment by patrickdillon** [comment]
> /cc @jstuever @patrickdillon
> 
> Would one of you mind reviewing this PR, please?
/assign

File: pkg/operator/migration_platform_status/controller.go [hunk 1]
```diff
@@ -75,41 +72,72 @@ func (c MigrationAWSStatusController) sync(ctx context.Context, syncCtx factory.
...
+	}
+
+	if currentInfra.Status.PlatformStatus.AWS.Region != "" {
+		// region is already set, so no need to migrate
+		return nil
 	}
 
 	cc, err := loadClusterConfig(ctx, c.configMapClient)
 	if err != nil {
-		syncCtx.Recorder().Warningf("MigrationAWSStatusController", "Unable to load the cluster-config-v1")
```
Comment (deads2k) [inline pkg/operator/migration_platform_status/controller.go:87]: this still seems event worthy

**Comment by deads2k** [comment]
notice that the old condition is still present: `MigrationAWSStatusControllerDegraded`.  You need the stale condition controller from library-go to remove it.  See examples in kube-scheduler operator.

File: pkg/operator/migration_platform_status/controller.go (hunk 1 above)
Comment (staebler) [reply pkg/operator/migration_platform_status/controller.go:87]: This error will be recorded as an event by the caller of `migratePlatformSpecificFields`.

File: pkg/operator/migration_platform_status/controller.go [hunk 2]
```diff
@@ -75,41 +72,72 @@ func (c MigrationAWSStatusController) sync(ctx context.Context, syncCtx factory.
...
+		syncCtx.Recorder().Warningf("MigrationPlatformStatusController", message)
+		return fmt.Errorf(message)
+	}
+
+	if err := c.migratePlatformSpecificFields(ctx, currentInfra); err != nil {
+		syncCtx.Recorder().Warningf("MigrationPlatformStatusController", err.Error())
+		return err
+	}
+
+	if reflect.DeepEqual(obji.Status.PlatformStatus, currentInfra.Status.PlatformStatus) {
```
Comment (soltysh) [inline pkg/operator/migration_platform_status/controller.go:92]: `equality.Semantic.DeepEqual` is preferred for our API types

**Comment by staebler** [comment]
e5c6695c...ddba24bb
* Add stale condition controller to remove `MigrationAWSStatusControllerDegraded` condition.

**Comment by staebler** [comment]
ddba24bb...1e5543d5
* Replaced use of `reflect.DeepEqual` with `equality.Semantic.DeepEqual`.

**Comment by staebler** [comment]
> notice that the old condition is still present: `MigrationAWSStatusControllerDegraded`. You need the stale condition controller from library-go to remove it. See examples in kube-scheduler operator.

@deads2k Would you mind giving this another look to see if I have satisfied your concerns?

**Comment by staebler** [comment]
@patrickdillon Could you look over this again with a focus on the recent changes made to add the stale condition controller?

**Comment by deads2k** [comment]
/approve



> @patrickdillon Could you look over this again with a focus on the recent changes made to add the stale condition controller?

if you could check the conditions this time, that would be great.

**Comment by patrickdillon** [comment]
> @patrickdillon Could you look over this again with a focus on the recent changes made to add the stale condition controller?

> if you could check the conditions this time, that would be great.

Sure. I don't have a lot of experience with this, so I will spend some time kicking the tires today.

**Comment by staebler** [comment]
> /approve
> 
> > @patrickdillon Could you look over this again with a focus on the recent changes made to add the stale condition controller?
> 
> if you could check the conditions this time, that would be great.

@patrickdillon These are the steps that I took to test that the stale `MigrationAWSStatusControllerDegraded` condition is removed.

1. Install an OpenShift 4.7.0-fc3 cluster.
2. Remove the .platform.aws.region field from the install-config in the kube-system/cluster-config-v1 configmap.
3. Remove the .status.platformStatus field from the infrastructure resource.
Start a proxy in another console via `oc proxy`.
```
$ oc get --raw /apis/config.openshift.io/v1/infrastructures/cluster | jq 'del(.status.platformStatus)' > infrastructure.json 
$ curl -k -H "Content-Type: application/json" -X PUT --data-binary @infrastructure-new.json http://127.0.0.1:8001/apis/config.openshift.io/v1/infrastructures/cluster/status
```
4. Verify that the config-operator clusteroperator is degraded with MigrationAWSStatusControllerDegraded.
```
$ oc get co config-operator -oyaml | yq -y '.status.conditions[] | select(.type=="Degraded")'
    message: |-
      MigrationAWSStatusControllerDegraded: empty region set in cluster-config-v1
      KubeCloudConfigControllerDegraded: invalid platform, expected to be AWS
    reason: AsExpected
    status: "False"
    type: Degraded
```
5. Upgrade to an OpenShift version with these PR changes. I built the image that I used with cluster-bot.
```
oc adm upgrade --to-image=registry.svc.ci.openshift.org/ci-ln-n7rhnct/release:latest --force --allow-explicit-upgrade
```
6. Wait for the config-operator clusteroperator to complete its upgrade.
7. Verify that the config-operator clusteroperator is still degraded but now with MigrationPlatformStatusControllerDegraded instead of MigrationAWSStatusControllerDegraded.
```
$ oc get co config-operator -oyaml | yq -y '.status.conditions[] | select(.type=="Degraded")'
    message: |-
      KubeCloudConfigControllerDegraded: invalid platform, expected to be AWS
      MigrationPlatformStatusControllerDegraded: empty region set in cluster-config-v1
    reason: KubeCloudConfigController_SyncError::MigrationPlatformStatusController_SyncError
    status: "True"
    type: Degraded
```
8. Remove the .platform.aws.region field back to the install-config in the kube-system/cluster-config-v1 configmap.
9. Verify that the config-operator clusteroperator is no longer degraded.
```
$ oc get co config-operator -oyaml | yq -y '.status.conditions[] | select(.type=="Degraded")'
    reason: AsExpected
    status: "False"
    type: Degraded
```

**Pull request merged**